### PR TITLE
Add CanvasCast Extension

### PR DESCRIPTION
**Pull request by YodaLightsabr**
## Description

Add Canvas LMS extension.

## Type of change

New extension

## Screencast

There isn't a lot to show because it's one command, but here it is. I also can't show the actual data because it has personal information.

image

Courses load a user's courses from https://canvas.instructure.com/doc/api/courses.html
Assignments load a user's TODO items from https://canvas.instructure.com/doc/api/users.html
Announcements load a user's announcements from https://canvas.instructure.com/doc/api/announcements.html

All colors are linked to the color of the course.

## Checklist

- [x ] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x ] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x ] I checked that extension's icon has size of `512x512`

### CODE REVIEW

**Comment by maggie-j-liu** [comment]
hiii! this extension looks super cool; i think it would be nice to add a readme that explains how to get the access token, and also add a warning about the access token breaking canvas' tos as explained [here](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

**Comment by pernielsentikaer** [comment]
@YodaLightsabr Do you need help with anything?

**Comment by YodaLightsabr** [comment]
I'll have a few more test assignments added soon, I've just been really busy and haven't gotten around to it yet.

**Comment by pernielsentikaer** [comment]
Hi @YodaLightsabr 👋 

 Sure! Just DM on Slack when you're ready 🧑‍🎤

And as @maggie-j-liu said, there should be a README.md file telling about the extension (and how to get token) and if anything not is compliant with the terms it should be added too: https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration

**Comment by fe9lix** [comment]
Closing this PR due to inactivity – feel free to reopen later.

**Comment by YodaLightsabr** [comment]
Thanks, closing it reminded me I hadn't updated it. I think I've just about finished adding all of the suggestions.

**Comment by fe9lix** [comment]
@YodaLightsabr  Let me know if you want to reopen it here...

**Comment by YodaLightsabr** [comment]
Yeah, I think it's ready to be reopened.

**Comment by pernielsentikaer** [comment]
Hi @YodaLightsabr 

Let's reopen it and get it merged – we could call it “My Courses” instead, that would be a more sufficient title

**Comment by YodaLightsabr** [comment]
> Hi @YodaLightsabr
> 
> Let's reopen it and get it merged – we could call it “My Courses” instead, that would be a more sufficient title

Do you mean in the courses section of the command?

File: extensions/canvascast/package.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      "required": true,
+      "title": "Domain",
+      "placeholder": "canvas.instructure.com",
+      "description": "Domain for your canvas LMS. Do not include URL; just the domain."
+    }
+  ],
+  "commands": [
+    {
+      "name": "index",
+      "title": "Open Dashboard",
```
Comment (pernielsentikaer) [inline extensions/canvascast/package.json:29]:
```suggestion
      "title": "My Courses",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @YodaLightsabr 👋 

See my latest suggestion 😄

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Review by pernielsentikaer — APPROVED** [review]
Hi @YodaLightsabr 👋 

I'm going to accept it and then will my colleague do a final review before it's finally merged 💪

**Comment by eliotbehr** [comment]
@YodaLightsabr 

This is a very useful extension. I do have one suggestion: Can you please make the extension only show courses which we have bookmarked in Canvas. For students with many old, but still Canvas-enrolled courses, it takes up the entire CanvasCast "Enrolled courses" thing. E.g., I can't see any course with a letter after "B". By only showing "Bookmarked courses", it would allow us to easily access the courses we want as most accessible and hopefully prevent these past-enrolled courses from taking up the entire CanvasCast pop-up.

**Comment by YodaLightsabr** [comment]
Oh yeah, I can totally relate to that. I have a ton of old courses and it wasn't until recently where they were dropped from my account. I'll take a look at this soon.

**Comment by eliotbehr** [comment]
> Oh yeah, I can totally relate to that. I have a ton of old courses and it wasn't until recently where they were dropped from my account. I'll take a look at this soon.

Thanks so much for looking into this! For colleges like mine that are too lazy to drop them it would be super helpful haha

**Comment by YodaLightsabr** [comment]
@eliotbehr if you could test something for me it would be a big help. Could you sign into your canvas and then GET `https://<YOUR_CANVAS_DOMAIN>/api/v1/courses` and compare it to the response of `https://<YOUR_CANVAS_DOMAIN>/api/v1/courses?enrollment_state=active`. I just want to know if adding the query param excludes your old courses.

**Comment by eliotbehr** [comment]
@YodaLightsabr It did change it very slightly—with enrollment_state=active it removed one course that the school seemed to have explicitly unenrolled me, but all the old courses where I'm still "enrolled" are there. I also noticed that this API query seems to limit to retrieving 10 courses maximum.

Could this be a possible option to query?: api/v1/users/self/favorites/courses

**Comment by YodaLightsabr** [comment]
Oh, I didn't see that. As long as it returns the same Course object, it should be an easy swap.

**Comment by eliotbehr** [comment]
@YodaLightsabr 
Hey Yoda! Thank you so much for working on these changes! I saw that the extension was recently updated, and I re-installed it but I am still seeing the same old courses. Is there anyway I could help troubleshoot this—perhaps is it because of the API still?

**Comment by YodaLightsabr** [comment]
Hmmm, that shouldn't be happening. Could you tell me the API outputs of the courses endpoint, and then the favorites endpoint, and then what you want to see in the extension? You can hide private course details, it would just be helpful for debugging.

**Comment by gronk-droid** [comment]
@YodaLightsabr 
Hey there! I'm having the same issue with favorites here. My unfavorited courses in Canvas are still showing in the command.

<div class="grid-containter" style="display: grid; grid-template-columns: auto auto">
<div class="grid-item" style="text-align: center">
Screen Shot 2022-02-05 at 3 02 31 PM
</div>
<div class="grid-item">
Screen Shot 2022-02-05 at 3 02 46 PM
</div>
</div>

**Comment by YodaLightsabr** [comment]
Ok, I think I've figured out how to only use the courses on the Dashboard and sort them in that order. I'll get started on implementing it.

**Pull request merged**